airflow/providers/amazon/aws/hooks/s3.py

```diff
@@ -36,24 +37,48 @@
 def provide_bucket_name(func):
     """
     Function decorator that provides a bucket name taken from the connection
-    in case no bucket name has been passed to the function and, if available, also no key has been passed.
+    in case no bucket name has been passed to the function.
     """
 
     @wraps(func)
     def wrapper(*args, **kwargs):
-        func_params = func.__code__.co_varnames
+        bound_args = signature(func).bind(*args, **kwargs)
```

Comment:
We should get `signature(func)` outside the wrapper as it's probably "relatively" expensive and it won't change form call-to-call.

Reply:
Good point!